### PR TITLE
Optimizers pass in tape mode

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

It's the optimizers turn to begin passing in tape mode.

**Description of the Change:**

* Adds fixture to `test_optimizer` so that it is run in tape mode. No real modifications needed here.

* Adds fixture to `test_optimizer_qng` so that it is run in tape mode.

* Modifies `qml.metric_tensor()` so that it works with old QNodes, tape QNodes, and `ExpvalCost`

* Modifies `ExpvalCost` to allow the generation of the metric tensor in tape mode QNodes.

**Benefits:** As above.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/vqe/vqe.py [hunk 1]
```diff
@@ -490,30 +490,12 @@ def __init__(
...
-
-                self._qnode_for_metric_tensor_in_tape_mode = qnode_for_metric_tensor_in_tape_mode
-            finally:
-                qml.enable_tape()
-
         self._optimize = optimize
 
+        self.qnodes = qml.map(
+            ansatz, observables, device, interface=interface, diff_method=diff_method, **kwargs
+        )
```
Comment (trbromley) [inline pennylane/vqe/vqe.py:497]: My only worry of moving this here is when `self._optimize is True`, then `self.qnodes` will be visible to the user but disconnected from the actual QNode used to evaluate.

**Review by trbromley — APPROVED** [review]
Thanks @josh146 !

**Comment by josh146** [comment]
I think the codecov check is only failing because I deleted so many lines, that the coverage percentage decreased...

**Review by albi3ro — COMMENTED** [review]
Looks great!

Can't see anything worth commenting on, but I am curious why `metric_tensor` got moved from a method within a qnode to a function that takes a qnode as an input.  Was there a specific thing that fixed, or is it part of a certain design philosophy?

**Comment by josh146** [comment]
> but I am curious why metric_tensor got moved from a method within a qnode to a function that takes a qnode as an input. Was there a specific thing that fixed, or is it part of a certain design philosophy?

A design philosophy 🙂 We've found lately that the pattern

```python
>>> qnode.something(options, *args, **kwargs)
```

hampers visibility, and has side effects if it uses or changes the QNode state.

Instead, we've been slowly moving to

```python
>>> qml.something(qnode, options)(*args, **kwargs)
```

Another advantage of this approach is that it means the `qml.something()` function can apply to _more_ than just QNodes if needed, as in this case where it applies to `ExpvalCost`.

File: pennylane/tape/qnode.py [hunk 2]
```diff
@@ -901,6 +901,20 @@ def circuit(weights):
     array([[ 0.04867729, -0.00049502,  0.        ],
            [ 0.        ,  0.        ,  0.        ]])
     """
+    if _qnode.__class__.__name__ == "ExpvalCost":
+        if _qnode._multiple_devices:
```
Comment (josh146) [inline pennylane/tape/qnode.py:905]:
```suggestion
        if _qnode._multiple_devices:  # pylint: disable=protected-access
```

**Pull request merged**